File: src/main/java/application/MainController.java
package application;

import javafx.AngleDisplay;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.geometry.Point2D;
import javafx.scene.layout.AnchorPane;
import math.Vectors;

import java.net.URL;
import java.util.ArrayList;
import java.util.Map;
import java.util.ResourceBundle;

public class MainController implements Initializable {

  @FXML
  private AnchorPane parent;

  @FXML
  private RayTabController rayTabController;
  @FXML
  private GeneralTabController generalTabController;
  @FXML
  private OpticsTabController opticsTabController;
  @FXML
  private AnimationTabController animationTabController;
  @FXML
  private Point2D movementDelta;

  @Override
  public void initialize(URL location, ResourceBundle resources) {
    rayTabController.initialize(parent);
    generalTabController.initialize(parent);
    opticsTabController.initialize(parent);
    animationTabController.initialize(parent);
    Storage.opticsTabController = opticsTabController;
    Storage.rayTabController = rayTabController;
    Storage.parent = parent;
    parent.setOnMousePressed(event -> {
      if (!Storage.isAnimating) movementDelta = new Point2D(event.getSceneX(), event.getSceneY());
    });
    parent.setOnMouseDragged(event -> {
      if (Storage.isAnimating) return;
      Point2D newOffset = new Point2D(event.getSceneX(), event.getSceneY()).subtract(movementDelta);
      Storage.setOffset(Storage.getOffset().add(newOffset));
      movementDelta = new Point2D(event.getSceneX(), event.getSceneY());
    });
    parent.setOnMouseMoved(event -> {
      Point2D coords = new Point2D(event.getSceneX(), event.getSceneY());
      ArrayList<Point2D> remove = new ArrayList<>();
      for (Map.Entry<Point2D, AngleDisplay> entry : Storage.intersectionPoints.entrySet()) {
        if (!parent.getChildren().contains(entry.getValue())) {
          remove.add(entry.getKey());
          continue;
        }
        if (Vectors.distanceSquared(entry.getKey(), coords) < 400) {
          entry.getValue().setVisible(true);
          entry.getValue().setLayoutX(event.getSceneX() + 7);
          entry.getValue().setLayoutY(event.getSceneY() + 7);
        } else {
          entry.getValue().setVisible(false);
        }
      }
      remove.forEach(point2D -> Storage.intersectionPoints.remove(point2D));
    });

  }
}

File: src/main/java/application/MainController.kt
package application

import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.layout.*
import java.net.URL
import java.util.*

class MainController : Initializable {
    @FXML
    private lateinit var parent: AnchorPane

    @FXML
    private lateinit var generalTabController: GeneralTabController

    override fun initialize(location: URL, resources: ResourceBundle?) {
        generalTabController.initialize(parent)
        Storage.generalTabController = generalTabController
        Storage.parent = parent
        parent.background = Background(
            BackgroundImage(
                Image(Main::class.java.getResourceAsStream("/susco.png")),
                BackgroundRepeat.REPEAT,
                BackgroundRepeat.REPEAT,
                BackgroundPosition.DEFAULT,
                BackgroundSize.DEFAULT
            )
        )
    }
}
